contents/javascript/Javascript-modules.md

```diff
@@ -0,0 +1,222 @@
+
+To include the module in `index.js`, we use the `<script>` tag in HTML to import it.
+
+```html
+<!-- index.html -->
+
+<script src="./anExampleModule.js" />
+<script src="./index.js" />
+```
+
+```js
+// anExampleModule.js
+
+var anExampleModule = (function() {
+    var variableOne = 1;
+    var variableTwo = 2;
+
+    var sumOfVariable = function() {
+        return variableOne + variableTwo;
+    }
+
+    // We return an object with functions or variables
+    // that we want to be exposed
+    return {
+        sumOfVariable: sumOfVariable
+    }
+})()
+```
+
+```js
+// index.js
+
+anExampleModule.sumOfVariable(); // 3
+```
+
+The result of importing the modules in the HTML will be as follow:
+```js
+// anExampleModule.js
+
+var anExampleModule = (function() {
+    var variableOne = 1;
+    var variableTwo = 2;
+
+    var sumOfVariable = function() {
+        return variableOne + variableTwo;
+    }
+
+    // We return an object with functions or variables
+    // that we want to be exposed
+    return {
+        sumOfVariable: sumOfVariable
+    }
+})()
+
+// index.js
+
+anExampleModule.sumOfVariable(); // 3
+```
+
+### CommonJS
+
+It is one of the most popular modules implementation before ES6 is rolled out because it is widely used by NodeJS. It still can be used without using NodeJS by using [bundler](https://medium.com/@gimenete/how-javascript-bundlers-work-1fc0d0caf2da) such as [Webpack](https://webpack.js.org/).
+
+Using CommonJS is quite similar to using IIFE, but instead of returning the functions or variable we want to expose, instead we assign it to `module.export`. To import the module, we can use `require` instead of using `<script>` tag in HTML.
+
+See the following snippets for example of using CommonJS:
+
+```js
+// anExampleModule.js
+
+var variableOne = 1;
+var variableTwo = 2;
+
+sumOfVariable = function() {
+    return variableOne + variableTwo;
+}
+
+module.exports = {
+  sumOfVariable: sumOfVariable,
+};
+```
+
+```js
+// index.js
+
+var anExampleModule = require('./anExampleModule.js');
+
+anExampleModule.sumOfVariable(); // 3
+```
+
+### EcmaScript2015 / ES6 modules
+
+In 2015, JavaScript introduced built-in modules in ES6. It uses the syntax `import` and `export`.
+
+```js
+// anExampleModule.js
+
+var variableOne = 1;
+var variableTwo = 2;
+
+export sumOfVariable() {
+    return variableOne + variableTwo;
+}
+```
+
+```js
+// index.js
+
+import * as anExampleModule from './anExampleModule.js';
+
+anExampleModule.sumOfVariable(); // 3
+```
+
+### Which to use?
+
+While ES6 modules is officially the most recent JavaScript standard rolled out, it is still not widely supported by all browsers. It needs [transpiler](https://scotch.io/tutorials/javascript-transpilers-what-they-are-why-we-need-them) such as [Babel](https://babeljs.io/) for unsupported browser. Similar to CommonJS, ES6 modules also requires bundler such as Webpack.
```

Comment:
As per [given here](https://developers.google.com/web/fundamentals/primers/modules), it is supported in all major browsers. That means this article is going to be outdated pretty soon, if not outdated already. Perhaps explain ES6 system well and just list the other options with hyperlinks?

Reply:
Node.JS supports for ES6 modules is still an experimental feature, and afaik people is still using CommonJS in NodeJS. There are also big numbers of applications built on legacy module system. I think it will be better to explain all those 3 implementations or else reader might not understand if they are working with application using CommonJS and Module Pattern